File: cpp/cpp_pt/materials/dielectric.cpp
#include <cpp_pt/materials/dielectric.hpp>
#include <cpp_pt/math/details.hpp>

dielectric::dielectric(const colour &diffuse_in, double refractive_index_in) :
   diffuse(diffuse_in), refractive_index(refractive_index_in)
{}

auto dielectric::scatter(const ray &ray_in, const hit &hit_in, [[maybe_unused]] double u,
   [[maybe_unused]] double v) const -> scatter_data
{
   const vec unit_dir = normalise(ray_in.direction());
   const double ior_ratio = (hit_in.front_face) ? (1.0 / refractive_index) : refractive_index;
   const double cos_i = -dot(hit_in.normal, unit_dir);
   const double sin_t_squared = ior_ratio * ior_ratio * (1.0 - cos_i * cos_i);

   if (sin_t_squared > 1.0) // total internal reflection
   {
      // clang-format off
      return scatter_data {
         .emission = colour{}, 
         .diffuse = diffuse, 
         .scattered_ray = ray(hit_in.position, reflect(hit_in.normal, unit_dir))
      };
      // clang-format on
   }

   if (random_double() < schlick(cos_i, ior_ratio))
   {
      // clang-format off
      return scatter_data {
         .emission = colour{}, 
         .diffuse = diffuse, 
         .scattered_ray = ray(hit_in.position, reflect(hit_in.normal, unit_dir))
      };
      // clang-format on
   }

   const double cos_t = std::sqrt(1.0 - sin_t_squared);
   vec refracted = ior_ratio * unit_dir + (ior_ratio * cos_i - cos_t) * hit_in.normal;

   // clang-format off
   return scatter_data {
      .emission = colour{}, 
      .diffuse = diffuse, 
      .scattered_ray = ray(hit_in.position, refracted)
   };
   // clang-format on
}

File: go/materials/dielectric.go
package materials

import (
	"go_pt/maths"
	"go_pt/tracer"
	"math"
	"math/rand"
)

type Dielectric struct {
   diffuseColour maths.Vec3
   refractiveIndex float64
}

func (mat *Dielectric) Scatter(ray *tracer.Ray, hit *tracer.Hit, u, v float64) MaterialScatterData {
   unitDir := maths.Normalise(&ray.Direction) 

   var iorRatio float64
   if hit.FrontFace {
      iorRatio = 1.0 / mat.refractiveIndex
   } else {
      iorRatio = mat.refractiveIndex
   }

   cosI := - maths.Dot(&hit.Normal, &unitDir)
   sinTSquared := iorRatio * iorRatio * (1.0 * cosI * cosI)

   if sinTSquared > 1.0 || rand.Float64() < maths.Schlick(cosI, iorRatio) {
      return MaterialScatterData{
         Emission: maths.Vec3{},
         Diffuse: mat.diffuseColour,
         ScatteredRay: tracer.Ray{
            Origin: hit.Position, 
            Direction: maths.Reflect(&hit.Normal, &unitDir)}}
   }

   cosT := math.Sqrt(1.0 - sinTSquared)
   refracted := maths.AddCpy(
      maths.MultScalar(&unitDir, iorRatio),
      maths.MultScalar(&hit.Normal, iorRatio * cosI - cosT))

   return MaterialScatterData{
      Emission: maths.Vec3{},
      Diffuse: mat.diffuseColour,
      ScatteredRay: tracer.Ray{
         Origin: hit.Position, 
         Direction: maths.Reflect(&hit.Normal, &refracted)}}
}
